### PR TITLE
Fix the logic of fallback to 16 bits index buffer in Mesh.cpp, also fix memory leak if vertex count is less than 65536 on the platform which doesn't support ELEMENT_INDEX_UINT.

### PR DESCRIPTION
**Pull request by dumganhar**
-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/base/RefVector.h [hunk 1]
```diff
@@ -189,6 +202,25 @@ class RefVector {
...
+            //            CC_LOG_INFO("In the move assignment operator!");
+            clear();
+            _data = std::move(other);
+            // NOTE: The reference count of T in ccstd::vector may be 0 so we need to retain all elements in ccstd::vector.
+            addRefForAllObjects();
+        }
+        return *this;
+    }
+
+    RefVector<T> &operator=(std::initializer_list<T> list) {
```
Comment (dumganhar) [inline native/cocos/base/RefVector.h:216]: It's for the usage of `RefVector<T>  x = {}`

File: native/cocos/3d/assets/Mesh.cpp [hunk 2]
```diff
@@ -309,21 +309,27 @@ void Mesh::initialize() {
...
-
-                //            ib = ccnew (getIndexStrideCtor(idxView.stride))(buffer, idxView.offset, idxView.count);
                 if (idxView.stride != dstStride) {
-                    //cjh  need in c++?              ib = getIndexStrideCtor(dstStride).from(ib);
+                    uint32_t ib16BitLength = idxView.length >> 1;
+                    uint16_t *ib16Bit = static_cast<uint16_t*>(CC_MALLOC(ib16BitLength));
+                    const uint32_t *ib32Bit = reinterpret_cast<const uint32_t*>(ib);
+                    for (uint32_t j = 0, len = idxView.count; j < len; ++j) {
+                        ib16Bit[j] = ib32Bit[j];
+                    }
```
Comment (dumganhar) [inline native/cocos/3d/assets/Mesh.cpp:321]: Convert to 16 bits index format.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Review by minggo — APPROVED** [review]

**Comment by dumganhar** [comment]
Testcase auto test failed. I don't know why. It seems it could not work now.

**Pull request merged**